### PR TITLE
Make jl_binding_t into a first-class object

### PR DESCRIPTION
**Pull request by apaz-cli**
There are two steps to this, and this is step 1.

1. Make `jl_binding_t` a first-class object (`Core.Binding`)
2. Combine `jl_binding_t` and `jl_globalref_t`, and rip out all the special-casing in the garbage collector and everywhere else.

### CODE REVIEW

**Review by vtjnash — APPROVED** [review]

**Comment by apaz-cli** [comment]
I also have a concern that it's a little bit of an api break.

In `doc/src/manual/embedding.md`, we recommend that people do:
```c
jl_binding_t *bp = jl_get_binding_wr(jl_main_module, jl_symbol("var"), 1);
```
If anybody actually took that advice, then combining `jl_binding_t` and `jl_globalref_t` could break some existing code. The embedding api would have to change somewhat.

Also, on CI:
LoadError(at "compiler/compiler.jl" line 3: LoadError(at "essentials.jl" line 578: LoadError(at "essentials.jl" line 578: TypeError(func=:new, context="", expected=Core.Binding, got=0x0000000000000000))))

This is a result of:
https://github.com/JuliaLang/julia/pull/47592/files#diff-882927c6e612596e22406ae0d06adcee88a9ec05e8b61ad81b48942e2cb266e9L2688-R2694

I'm not sure what should be here instead to make it nullable, but give the correct type in the interpreter.

File: src/jltypes.c [hunk 1]
```diff
@@ -2682,10 +2682,16 @@ void jl_init_types(void) JL_GC_DISABLED
...
+                        jl_perm_symsvec(6, "name", "value", "globalref", "owner", "ty", "flags"),
+                        jl_svec(6, jl_symbol_type, jl_any_type, jl_any_type, jl_module_type, jl_any_type, jl_uint8_type),
+                        jl_emptysvec, 0, 0, 1);
+
     jl_globalref_type =
         jl_new_datatype(jl_symbol("GlobalRef"), core, jl_any_type, jl_emptysvec,
                         jl_perm_symsvec(3, "mod", "name", "binding"),
-                        jl_svec(3, jl_module_type, jl_symbol_type, pointer_void),
+                        jl_svec(3, jl_module_type, jl_symbol_type, jl_binding_type),
                         jl_emptysvec, 0, 0, 3);
```
Comment (JeffBezanson) [inline src/jltypes.c:2695]:
```suggestion
                        jl_emptysvec, 0, 0, 2);
```
If the last field can be undefined then this needs to be 2.

File: src/jltypes.c [hunk 2]
```diff
@@ -2682,10 +2682,16 @@ void jl_init_types(void) JL_GC_DISABLED
 
     jl_value_t *pointer_void = jl_apply_type1((jl_value_t*)jl_pointer_type, (jl_value_t*)jl_nothing_type);
 
+    jl_binding_type =
+        jl_new_datatype(jl_symbol("Binding"), core, jl_any_type, jl_emptysvec,
+                        jl_perm_symsvec(6, "name", "value", "globalref", "owner", "ty", "flags"),
+                        jl_svec(6, jl_symbol_type, jl_any_type, jl_any_type, jl_module_type, jl_any_type, jl_uint8_type),
+                        jl_emptysvec, 0, 0, 1);
```
Comment (JeffBezanson) [inline src/jltypes.c:2689]:
```suggestion
                        jl_emptysvec, 0, 1, 1);
```

**Comment by JeffBezanson** [comment]
With this can we remove jl_gc_wb_binding and jl_gc_queue_binding?

**Comment by apaz-cli** [comment]
@JeffBezanson Yeah. The idea is that since there's a 1-1 correspondence between `jl_binding_t` and `jl_globalref_t` (although we do delay creating a binding for a globalref, so it may actually be a 1-none correspondance), they should be the same object. We can store the binding inside the globalref, and with that change we would rip out the functions that create or deal with bindings, and make them take the globalref.

File: src/jltypes.c [hunk 3]
```diff
@@ -2682,11 +2682,17 @@ void jl_init_types(void) JL_GC_DISABLED
 
     jl_value_t *pointer_void = jl_apply_type1((jl_value_t*)jl_pointer_type, (jl_value_t*)jl_nothing_type);
 
+    jl_binding_type =
+        jl_new_datatype(jl_symbol("Binding"), core, jl_any_type, jl_emptysvec,
+                        jl_perm_symsvec(6, "name", "value", "globalref", "owner", "ty", "flags"),
+                        jl_svec(6, jl_symbol_type, jl_any_type, jl_any_type, jl_module_type, jl_any_type, jl_uint8_type),
+                        jl_emptysvec, 0, 1, 1);
```
Comment (vtjnash) [inline src/jltypes.c:2689]:
```suggestion
                        jl_emptysvec, 0, 1, 1);
                        const static uint32_t binding_constfields[1]  = { 0x0001 }; // Set fields 1 as const
                        const static uint32_t binding_atomicfields[1] = { 0x0016 }; // Set fields 2, 3, 5 as atomic
                        jl_binding_type->name->constfields = binding_constfields;
                        jl_binding_type->name->atomicfields = binding_atomicfields;
```

File: src/jltypes.c [hunk 4]
```diff
@@ -2682,11 +2682,17 @@ void jl_init_types(void) JL_GC_DISABLED
...
+
     jl_globalref_type =
         jl_new_datatype(jl_symbol("GlobalRef"), core, jl_any_type, jl_emptysvec,
                         jl_perm_symsvec(3, "mod", "name", "binding"),
-                        jl_svec(3, jl_module_type, jl_symbol_type, pointer_void),
-                        jl_emptysvec, 0, 0, 3);
+                        jl_svec(3, jl_module_type, jl_symbol_type, jl_binding_type),
+                        jl_emptysvec, 0, 0, 2);
```
Comment (vtjnash) [inline src/jltypes.c:2695]:
```suggestion
                        jl_svec(3, jl_module_type, jl_symbol_type, pointer_void),
                        jl_emptysvec, 0, 0, 3);
```

File: src/jltypes.c [hunk 5]
```diff
@@ -2682,11 +2682,17 @@ void jl_init_types(void) JL_GC_DISABLED
 
     jl_value_t *pointer_void = jl_apply_type1((jl_value_t*)jl_pointer_type, (jl_value_t*)jl_nothing_type);
 
+    jl_binding_type =
+        jl_new_datatype(jl_symbol("Binding"), core, jl_any_type, jl_emptysvec,
+                        jl_perm_symsvec(6, "name", "value", "globalref", "owner", "ty", "flags"),
+                        jl_svec(6, jl_symbol_type, jl_any_type, jl_any_type, jl_module_type, jl_any_type, jl_uint8_type),
```
Comment (vtjnash) [inline src/jltypes.c:2688]:
```suggestion
                        jl_svec(6, jl_symbol_type, jl_any_type, jl_any_type/*jl_binding_type*/, jl_module_type, jl_any_type, jl_uint8_type),
```

**Pull request merged**